File: locus-api-android/src/main/java/locus/api/android/features/mapProvider/data/MapTileResponse.java
package locus.api.android.features.mapProvider.data;

import android.graphics.Bitmap;

import java.io.IOException;

import locus.api.android.utils.Utils;
import locus.api.android.utils.UtilsBitmap;
import locus.api.objects.Storable;
import locus.api.utils.DataReaderBigEndian;
import locus.api.utils.DataWriterBigEndian;

public class MapTileResponse extends Storable {

	public static final int CODE_UNKNOWN 			= 0;
	public static final int CODE_VALID				= 1;

	public static final int CODE_INVALID_REQUEST	= 2;
	public static final int CODE_NOT_EXISTS			= 3;
	public static final int CODE_INTERNAL_ERROR		= 4;
	
	/*
	 * Result code that indicate result of whole operation
	 */
	private int mResultCode;
	/*
	 * Image itself
	 */
	private Bitmap mImage;
	
	// BASIC CONSTRUCTORS
	
	public MapTileResponse() {
		super();
	}
	
	public MapTileResponse(byte[] data) throws IOException {
		super(data);
	}
	
	// GET & SET METHODS
	
	public int getResultCode() {
		return mResultCode;
	}

	public void setResultCode(int resultCode) {
		this.mResultCode = resultCode;
	}

	public Bitmap getImage() {
		return mImage;
	}

	public void setImage(Bitmap image) {
		this.mImage = image;
	}	
	
	/**************************************************/
	/*                 STORABLE PART                  */
	/**************************************************/
	
	@Override
	protected int getVersion() {
		return 0;
	}

	@Override
	public void reset() {
		mResultCode = CODE_UNKNOWN;
		mImage = null;
	}

	@Override
	protected void readObject(int version, DataReaderBigEndian dr)
			throws IOException {
		mResultCode = dr.readInt();
		
		// icon
		int size = 0;
		if ((size = dr.readInt()) > 0) {
			byte[] data = dr.readBytes(size);
			mImage = UtilsBitmap.getBitmap(data);
		} else {
			mImage = null;
		}
	}

	@Override
	protected void writeObject(DataWriterBigEndian dw) throws IOException {
		dw.writeInt(mResultCode);
		
		// icon
		if (mImage == null) {
			dw.writeInt(0);
		} else {
			byte[] data = UtilsBitmap.getBitmap(mImage, Bitmap.CompressFormat.PNG);
			if (data == null || data.length == 0) {
				dw.writeInt(0);
			} else {
				dw.writeInt(data.length);
				dw.write(data);
			}
		}
	}
}

File: locus-api-android/src/main/java/locus/api/android/features/mapProvider/data/MapTileResponse.kt
package locus.api.android.features.mapProvider.data

import android.graphics.Bitmap

import locus.api.android.utils.UtilsBitmap
import locus.api.objects.Storable
import locus.api.utils.DataReaderBigEndian
import locus.api.utils.DataWriterBigEndian

class MapTileResponse : Storable() {

    /*
     * Result code that indicate result of whole operation
     */
    var resultCode: Int = CODE_UNKNOWN
    /*
     * Image itself
     */
    var image: Bitmap? = null

    //*************************************************
    // STORABLE
    //*************************************************

    override fun getVersion(): Int {
        return 0
    }

    override fun readObject(version: Int, dr: DataReaderBigEndian) {
        resultCode = dr.readInt()

        // icon
        val size = dr.readInt()
        if (size > 0) {
            val data = dr.readBytes(size)
            image = UtilsBitmap.getBitmap(data)
        } else {
            image = null
        }
    }

    override fun writeObject(dw: DataWriterBigEndian) {
        dw.writeInt(resultCode)

        // icon
        if (image == null) {
            dw.writeInt(0)
        } else {
            val data = UtilsBitmap.getBitmap(image!!, Bitmap.CompressFormat.PNG)
            if (data == null || data.isEmpty()) {
                dw.writeInt(0)
            } else {
                dw.writeInt(data.size)
                dw.write(data)
            }
        }
    }

    companion object {

        const val CODE_UNKNOWN = 0
        const val CODE_VALID = 1

        const val CODE_INVALID_REQUEST = 2
        const val CODE_NOT_EXISTS = 3
        const val CODE_INTERNAL_ERROR = 4
    }
}
